scrapy/core/engine.py

```diff
@@ -148,92 +155,88 @@ def _needs_backout(self, spider: Spider) -> bool:
+    def _next_request_from_scheduler(self) -> Optional[Deferred]:
         assert self.slot is not None  # typing
 
         request = self.slot.scheduler.next_request()
         if request is None:
             return None
 
-        d = self._download(request, spider)
-        d.addBoth(self._handle_downloader_output, request, spider)
+        d = self._download(request)
+        d.addBoth(self._handle_downloader_output, request)
         d.addErrback(lambda f: logger.info('Error while handling downloader output',
                                            exc_info=failure_to_exc_info(f),
-                                           extra={'spider': spider}))
+                                           extra={'spider': self.spider}))
         d.addBoth(lambda _: self.slot.remove_request(request))
         d.addErrback(lambda f: logger.info('Error while removing request from slot',
                                            exc_info=failure_to_exc_info(f),
-                                           extra={'spider': spider}))
+                                           extra={'spider': self.spider}))
         d.addBoth(lambda _: self.slot.nextcall.schedule())
         d.addErrback(lambda f: logger.info('Error while scheduling new request',
                                            exc_info=failure_to_exc_info(f),
-                                           extra={'spider': spider}))
+                                           extra={'spider': self.spider}))
         return d
 
     def _handle_downloader_output(
-        self, result: Union[Request, Response, Failure], request: Request, spider: Spider
+        self, result: Union[Request, Response, Failure], request: Request
     ) -> Optional[Deferred]:
+        assert self.spider is not None  # typing
+
         if not isinstance(result, (Request, Response, Failure)):
             raise TypeError(f"Incorrect type: expected Request, Response or Failure, got {type(result)}: {result!r}")
+
         # downloader middleware can return requests (for example, redirects)
         if isinstance(result, Request):
-            self.crawl(result, spider)
+            self.crawl(result, self.spider)
             return None
-        d = self.scraper.enqueue_scrape(result, request, spider)
-        d.addErrback(lambda f: logger.error('Error while enqueuing downloader output',
-                                            exc_info=failure_to_exc_info(f),
-                                            extra={'spider': spider}))
+
+        d = self.scraper.enqueue_scrape(result, request, self.spider)
+        d.addErrback(
+            lambda f: logger.error(
+                'Error while enqueuing downloader output',
+                exc_info=failure_to_exc_info(f),
+                extra={'spider': self.spider},
+            )
+        )
         return d
 
     def spider_is_idle(self, spider: Spider) -> bool:
+        assert self.slot is not None  # typing
```

Comment:
Since this is a public method, we cannot control whether or not `self.slot` is or not `None` by the time this method is called, so I don’t think `assert` is the way here, since we should only use asserts if the code would work the same after removing them. I would raise an exception instead.

Same for `crawl` and `schedule` below.

Reply:
Makes sense, thanks